## Entropy

In [1]:
import numpy as np 
from math import log
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
#only 2 class
def E(a,b):
    if a == 0 or b ==0: #有一個是0，另一個必定是
0        return 0
    probs_a = a / (a+b)
    probs_b = b / (a+b)
    #Compute entropy
    ent = -(probs_a * log(probs_a,2) + probs_b * log(probs_b,2))
    
    return ent


In [3]:
S=E(9,5)
S

0.9402859586706309

### Outlook Gain

In [9]:
S-(5/14*E(3,2)+4/14*E(4,0)+5/14*E(2,3)) #所有因素結果是yes的情況(E)，分數為所有情況下符合當種情況的可能

0.2467498197744391

### Temp Gain

In [10]:
S-(4/14*E(2,2)+6/14*E(4,2)+4/14*E(3,1)) #E(yes,no)->單一種元素分析

0.029222565658954647

### HumilityGain

In [11]:
S- (7/14*E(3,4)+7/14*E(6,1))

0.15183550136234136

### Windy Gain

In [12]:
S-(8/14*E(6,2)+6/14*E(3,3))

0.04812703040826927

## Exercise :
    Outlook= rain 
    試著計算Temp, Humidity, Windy 的Information Gain 並畫出在Rainy下的樹狀結構
 
    
   ![image.png](attachment:image.png)

In [13]:
E(3,2)

0.9709505944546686

In [11]:
# temp gain
E(3,2) - (2/5*E(2,0)+2/5*E(1,1)+1/5*E(1,0))

0.5709505944546686

In [15]:
# humility gain
E(3,2) - (3/5*E(3,0)+2/5*E(2,0))

0.9709505944546686

In [13]:
# windy gain
E(3,2) - (3/5*E(2,1)+2/5*E(1,1))

0.01997309402197489

In [4]:
def entropy(labels):
    n_labels = len(labels)

    if n_labels <= 1:
        return 0

    value,counts = np.unique(labels, return_counts=True) #unique計算此元素出現的次數
    print(value,counts)
    probs = counts / n_labels

    ent = 0.
    # Compute entropy
    for i in probs:
        ent -= i * log(i,2)

    return ent


labels=['yes','yes','yes']
print(entropy(labels))

labels=['yes','no','yes','no']
print(entropy(labels))

labels=['high','low','high','high','low','normal','normal','normal']
print(entropy(labels))

labels=['high','high','high','high','high','normal','normal','normal']
print(entropy(labels))

['yes'] [3]
0.0
['no' 'yes'] [2 2]
1.0
['high' 'low' 'normal'] [3 2 3]
1.5612781244591325
['high' 'normal'] [5 3]
0.9544340029249649


## ans
['yes'] [3] <br>
0.0<br>
['no' 'yes'] [2 2]<br>
1.0<br>
['high' 'low' 'normal'] [3 2 3]<br>
1.5612781244591325<br>
['high' 'normal'] [5 3]<br>
0.9544340029249649<br>


In [5]:
from scipy.stats import entropy

In [6]:
entropy([1/2, 1/2], base=2)

1.0

In [7]:
entropy([5/8, 3/8], base=2)

0.954434002924965

In [8]:
0.97 - (2/5*entropy([2,0], base=2)+2/5*entropy([1,1], base=2)+1/5*entropy([1,0], base=2))


0.57